### DISCUSSION TITLE
Does mmdet have a profile to see how long each phase takes?

### DISCUSSION THREAD

**Discussion by vansin** [open-mmlab/mmdetection#8992] [General]
mmdet有profile吗，查看各个阶段耗时的 ?
Does mmdet have a profile to see how long each phase takes ?

**Comment by zachcoleman** [comment]
MMCV has a `ProfilerHook` see [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/profiler.py)

It uses the PyTorch profiler to show GPU utilization, dataloader time, and time spent in each operation. Here's an example of putting it in the config:
```python
profiler = dict(
    type="ProfilerHook",
    by_epoch=False,
    profile_iters=10,
    on_trace_ready=dict(type="tb_trace", dir_name="hook_prof_logs"),
)
```

In the TensorBoard output (specified by  setting `type` to be`tb_trace`) you can see output like this:
![Screen Shot 2022-10-14 at 8 57 06 AM](https://user-images.githubusercontent.com/42484306/195878053-5ef4e5e0-376e-416f-9705-8932e970fce0.png)

Your other option is to just use Python profiling and `gprof2dot` to visualize them. This is messy and sometimes hard to visualize and doesn't correspond to all the backend C/C++ code being called or give you an idea of how the GPU is working:
![Screen Shot 2022-10-14 at 8 58 23 AM](https://user-images.githubusercontent.com/42484306/195878695-10ced1c6-d5aa-4ce8-8b40-f6ea71268270.png)
